Add QMK Compile Context Sensitivity

cfbender opened the pull request:
<!--- Provide a general summary of your changes in the title above. -->

<!--- This template is entirely optional and can be removed, but is here to help both you and us. -->
<!--- Anything on lines wrapped in comments like these will not show up in the final text. -->

## Description

Adds context sensitivity to `qmk compile`. Allows the command to be run within a keyboard directory, given no other CLI arguments, and will compile the default keymap for the given keyboard.

<!--- Describe your changes in detail here. -->

## Types of Changes

<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply. -->
- [ ] Core
- [ ] Bugfix
- [ ] New feature
- [x] Enhancement/optimization
- [ ] Keyboard (addition or update)
- [ ] Keymap/layout/userspace (addition or update)
- [ ] Documentation

## Issues Fixed or Closed by This PR
 
* Closes #6824 

## Checklist

<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

cfbender commented:
As I said in the issue this closes, this is some of the first python I've written in a while, and the first of this nature. Please let me know if there's anything I can do better, or if anything is bad practice!

Maxr1998 commented on lib/python/qmk/cli/compile.py:
```diff
@@ -28,6 +29,25 @@ def compile(cli):
+        # If the folder above is "/keyboards/"
+        if above_basename == "keyboards":
+            # Set the full path name and get back the full keyboard name
+            path = Path(os.path.normpath(os.path.join(cwd, directory[index-1])))
```
Shouldn't this just be `directory`?
```suggestion
            path = Path(os.path.normpath(os.path.join(cwd, directory)))
```

cfbender replied:
It walks up the CWD until it finds the keyboards folder, and then sets the path to be the one directly below it. If it was just `directory` I believe it wouldn't work

cfbender replied:
Your comment made me rethink that line though, and made me realize that it was not necessary. I have cleaned it up, though it still relies upon getting the qmk directory from `os.getcwd()`

Maxr1998 commented on lib/python/qmk/cli/compile.py:
```diff
@@ -28,6 +29,25 @@ def compile(cli):
+        if above_basename == "keyboards":
+            # Set the full path name and get back the full keyboard name
+            path = Path(os.path.normpath(os.path.join(cwd, directory[index-1])))
+            keyboard += str(path.relative_to(os.getcwd() + "/keyboards"))
```
Maybe you should also put the QMK directory into an environment variable (in `bin/qmk`), implicitly using the current working directory seems error-prone to me..

cfbender replied:
I agree, that was one part I thought may be wrong as well. I am happy to try that out if that's a better approach

skullydazed replied:
The `qmk` script always changes current directory to the top level of `qmk_firmware`. You can find the directory the user ran the script from in `os.environ['ORIG_CWD']`. When looking for paths within qmk_firmware it's cleanest to use relative paths.

Maxr1998 replied:
True, but imo that isn't very clean/obvious when looking at the code. Maybe a `QMK_DIR` environment variable being set in the `qmk` script would still be a good idea? This might be another PR though.

skullydazed replied:
I'm open to discussing that further, but at this point we're out of scope for this PR. Can we move this to a discord discussion or a new issue?

skullydazed commented on lib/python/qmk/cli/compile.py:
```diff
@@ -28,6 +29,24 @@ def compile(cli):
 
     """
+    # Set CWD as directory command was issued from
+    cwd = os.path.normpath(os.path.join(os.environ['ORIG_CWD']))
```
You should just use `os.environ['ORIG_CWD']` directly. The path was normalized at the time it was stored.

skullydazed commented on lib/python/qmk/cli/compile.py:
```diff
@@ -28,6 +29,24 @@ def compile(cli):
+    in_keyboard = False
+    keyboard = ""
+
+    # Loop through three levels up to check for keyboards root directory
```
I would probably approach this differently, something like:

```
keyboards_path = os.path.join(os.getcwd(), 'keyboards')
if os.environ['ORIG_CWD'].startswith(keyboards_path):
    localpath = os.environ['ORIG_CWD'][len(keyboards_path):]
    # Parse down from localpath here
```

cfbender replied:
Ah, I like this. I will test this tonight!

cfbender replied:
I just tested this, the main thing is that it doesn't fail quite as gracefully when running `qmk compile` in the `keyboards` directory. Maybe not a necessary feature, but I will leave that up to you

cfbender commented:
@skullydazed that's great to hear. Would you like me to start to tackle the community layouts part as well? That side I'm a little less familiar with, since I've never utilized that side of the project.

skullydazed commented:
If you're up for it that'd be great! In a way community layouts are the inverse of keyboards- you have a keymap name but the user will have to supply a keyboard.

cfbender commented:
Great, I should have some time to work on that today/tomorrow. Would you want that in this PR, or in a separate one?

skullydazed commented:
Your choice on that. If you want to mark this as a draft for now while you work on it that's fine, or you can make it another PR so we can merge this one sooner.

cfbender commented:
@skullydazed I pushed some changes that hopefully take care of the layout and keymap context sensitivity :). Let me know what you think!

skullydazed commented on docs/cli.md:
```diff
@@ -71,7 +71,7 @@ There are some limitations to the local CLI compared to the global CLI:
 ## `qmk compile`
 
-This command allows you to compile firmware from any directory. You can compile JSON exports from <https://config.qmk.fm> or compile keymaps in the repo.
+This command allows you to compile firmware from any directory. You can compile JSON exports from <https://config.qmk.fm>, compile keymaps in the repo, or compile the keyboard in the current working directory.
```
Oxford comma. :D

skullydazed commented on docs/cli.md:
````diff
@@ -85,6 +85,52 @@ qmk compile <configuratorExport.json>
+
+```
+$ cd ~/qmk_firmware/keyboards/clueboard/66/rev4 
+$ qmk compile --keymap 66_iso
````
This should be consistent with the other usages above. 

```suggestion
$ qmk compile -km 66_iso
```

skullydazed commented on docs/cli.md:
````diff
@@ -85,6 +85,52 @@ qmk compile <configuratorExport.json>
+
+Must be under `qmk_firmware/layouts/`, and in a keymap folder.
+```
+qmk compile --keyboard <keyboard_name>
````
```suggestion
qmk compile -kb <keyboard_name>
```

skullydazed commented on docs/cli.md:
````diff
@@ -85,6 +85,52 @@ qmk compile <configuratorExport.json>
+**Example**:
+```
+$ cd ~/qmk_firmware/layouts/community/60_ansi/mechmerlin-ansi
+$ qmk compile --keyboard dz60
````
```suggestion
$ qmk compile -kb dz60
```

skullydazed commented on lib/python/qmk/cli/compile.py:
```diff
@@ -44,9 +76,34 @@ def compile(cli):
 
     else:
-        cli.log.error('You must supply a configurator export or both `--keyboard` and `--keymap`.')
+        cli.log.error('You must supply a configurator export or both `--keyboard` and `--keymap`, or be a directory for a keyboard or keymap.')
```
```suggestion
        cli.log.error('You must supply a configurator export, or both `--keyboard` and `--keymap`, or be in a directory for a keyboard or keymap.')
```

alexweininger commented:
Don't forget to mark this with a Hacktoberfest label!

cfbender commented:
@alexweininger I believe it doesn't matter this year, but it's still not a bad idea!

Maxr1998 commented:
Afaik a label would be needed so that the tool on hacktoberfest.qmk.fm counts your contributions, but the site also says that it isn't necessary that the tool sees all your contributions to get a shirt. Anyway, this is kinda OT from me here, so feel free to delete this comment once you read it :D

skullydazed commented:
This should be compiling the skully keymap:

```
Zachs-MacBook-Pro:rev4 zwhite$ qmk config user
user.keyboard=clueboard/66/rev3
user.keymap=skully
Zachs-MacBook-Pro:rev4 zwhite$ pwd
/Users/zwhite/qmk_firmware/keyboards/clueboard/66/rev4
Zachs-MacBook-Pro:rev4 zwhite$ qmk compile
Ψ Compiling keymap with make clueboard/66/rev4:default
```

This should not be detected as a keyboard directory, since it doesn't have a `rules.mk`:

```
Zachs-MacBook-Pro:66 zwhite$ pwd
/Users/zwhite/qmk_firmware/keyboards/clueboard/66
Zachs-MacBook-Pro:66 zwhite$ qmk compile
Ψ Compiling keymap with make clueboard/66:default
```

I think this is a bug:

```
Zachs-MacBook-Pro:skully zwhite$ pwd
/Users/zwhite/qmk_firmware/layouts/community/66_ansi/skully
Zachs-MacBook-Pro:skully zwhite$ qmk config user
user.keyboard=clueboard/66/rev3
user.keymap=skully
Zachs-MacBook-Pro:skully zwhite$ qmk compile
<class 'UnboundLocalError'>
☒ local variable 'command' referenced before assignment
Traceback (most recent call last):
  File "/usr/local/Cellar/qmk/0.0.15/libexec/lib/python3.7/site-packages/milc.py", line 564, in __call__
    return self.__call__()
  File "/usr/local/Cellar/qmk/0.0.15/libexec/lib/python3.7/site-packages/milc.py", line 569, in __call__
    return self._entrypoint(self)
  File "/Users/zwhite/qmk_firmware/lib/python/qmk/cli/compile.py", line 109, in compile
    cli.log.info('Compiling keymap with %s\n\n', ' '.join(command))
UnboundLocalError: local variable 'command' referenced before assignment
```

cfbender commented:
@skullydazed good catch. I have hopefully fixed those issues, I forgot about a user config option. I currently have it where `user` config overrules `compile` config, but happy to switch that if you think the compile should take precedence.

For the layout compile, it looks like I had updated the docs wrong, and it was assuming the keyboard name was a filename since the wrong flag was passed. I have fixed that!

cfbender commented:
@skullydazed ah, okay I wasn't sure what should take precedence, and I assumed the config. Easy fix! and same for the other, I figured default was the solution but I can change that as well.

cfbender commented on lib/python/qmk/cli/compile.py:
```diff
@@ -41,12 +83,38 @@ def compile(cli):
+        keymap_path = qmk.path.keymap(keyboard)
+        # Check for global keymap config first
+        if user_keymap:
+            command = ['make', ':'.join((keyboard, user_keymap))]
```
These are great changes, thanks for catching this! I have this committed now

Erovia commented on lib/python/qmk/cli/compile.py:
```diff
@@ -28,6 +29,48 @@ def compile(cli):
+    if cli.config.compile.keyboard or cli.config.user.keyboard:
+        user_keyboard = cli.config.user.keyboard if cli.config.user.keyboard else cli.config.compile.keyboard
+    if cli.config.compile.keymap or cli.config.user.keymap:
+        if not in_layout:
```
And the same goes for the keymap:

```suggestion

```

cfbender replied:
This would not solve the comment from above here: https://github.com/qmk/qmk_firmware/pull/6884#pullrequestreview-306268697. 

Removing this check will set `user_keyboard` and `user_keymap` based on the config, which would remove the layout directory context

Erovia replied:
Hmm, you are right.
Imho, we should work out the following precedence order: command line > directory context > config

skullydazed replied:
> Imho, we should work out the following precedence order: command line > directory context > config

Yes, this is the correct precedence.

Erovia commented on lib/python/qmk/cli/compile.py:
```diff
@@ -28,6 +29,48 @@ def compile(cli):
+        user_keyboard = cli.config.user.keyboard if cli.config.user.keyboard else cli.config.compile.keyboard
+    if cli.config.compile.keymap or cli.config.user.keymap:
+        if not in_layout:
+            user_keymap = cli.config.user.keymap if cli.config.user.keymap else cli.config.compile.keymap 
```
```suggestion
        user_keymap = cli.config.compile.keymap if cli.config.compile.keymap else cli.config.user.keymap 
```

cfbender commented on lib/python/qmk/cli/compile.py:
```diff
@@ -28,6 +29,48 @@ def compile(cli):
+
+    # If user keyboard/keymap or compile keyboard/keymap are supplied, assign those
+    if cli.config.compile.keyboard or cli.config.user.keyboard:
+        user_keyboard = cli.config.user.keyboard if cli.config.user.keyboard else cli.config.compile.keyboard
```
@skullydazed your thoughts on this? I believe this is something I had flipped before, but not sure. I tend to agree that the arguments should take precedence

skullydazed replied:
It should be even easier than that since milc does the right thing for you.

```suggestion
        user_keyboard = cli.config.compile.keyboard
```

Erovia replied:
With https://github.com/qmk/qmk_firmware/pull/7470 merged, this should work now.

skullydazed replied:
Now that #7470 is merged if you can merge this suggestion and the other suggestion I made we can approve this PR and get it merged.

cfbender commented:
> So, I just noticed these and I think these are pretty crucial for a satisfactory user experience.
> 
> One more thing, you should rebase on master as a few trivial merge conflicts will need fixin and if you are already there, might as well squash the commits.

Hmmm.. I'm not sure I did the rebase correctly, was I supposed to force push after rebasing?

Erovia commented:
Well, it became a bit messy in here, but it's easily fixable.
Are you on QMK Discord by any chance? That way would be easier.

drashna commented:
Looks like travis is erroring out here.

skullydazed commented on lib/python/qmk/cli/compile.py:
```diff
@@ -28,6 +29,47 @@ def compile(cli):
+    # If user keyboard/keymap or compile keyboard/keymap are supplied, assign those
+    if cli.config.compile.keyboard or cli.config.user.keyboard:
+        user_keyboard = cli.config.user.keyboard if cli.config.user.keyboard else cli.config.compile.keyboard
+    if cli.config.compile.keymap or cli.config.user.keymap:
```
```suggestion
    if cli.config.compile.keymap:
```

Erovia replied:
Note, this will only work correctly after https://github.com/qmk/qmk_firmware/pull/7470, as that adds inheritance handling with the dot notation.

cfbender commented:
Hi guys! Sorry it's been a hectic couple of weeks for me, I will get those changes Incorporated today!

skullydazed commented on docs/cli.md:
````diff
@@ -95,7 +95,54 @@ qmk compile <configuratorExport.json>
+...
+```
+
+## `qmk cformat`
````
```suggestion
## `qmk flash`
```

cfbender commented:
Okay hopefully that clears it up. I assume that milc will also do the same for the keymap, so I made that change as well. not sure why github is still showing requested changes, but I think we should be good now

Erovia commented:
@cfbender thanks for your work! It's on my priority to-do list to check your new commits out.

Erovia commented:
@cfbender thank you for your contribution!

cfbender commented:
@Erovia @skullydazed thanks for all of your help! Was a lot of fun, hope to work more on this project in the future!
